src/dbnode/storage/index/compaction/plan.go

```diff
@@ -0,0 +1,240 @@
+
+package compaction
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/storage/index/segments"
+)
+
+var (
+	errMutableCompactionAgeNegative = errors.New("mutable compaction age must be postive")
+	errSizeBucketsUndefined         = errors.New("size buckets are undefined")
+	errMaxImmutableCompactionSize   = errors.New("max immutable compaction size must be positive")
+)
+
+var (
+	DefaultSizeBuckets = []SizeBucket{ // i.e. tiers for compaction [0, 262K), [262K, 524K), [524K, 1M), [1M, 2M), [2M, 8M)
+		SizeBucket{
+			MinSizeInclusive: 0,
+			MaxSizeExclusive: 1 << 18,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 18,
+			MaxSizeExclusive: 1 << 19,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 19,
+			MaxSizeExclusive: 1 << 20,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 20,
+			MaxSizeExclusive: 1 << 21,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 21,
+			MaxSizeExclusive: 1 << 23,
+		},
+	}
+
+	DefaultOptions = PlannerOptions{
+		MaxImmutableCompactionSize: 1 << 22,                            // ~4M
+		MaxMutableSegmentSize:      1 << 16,                            // 64K
+		MutableCompactionAge:       15 * time.Second,                   // any mutable segment 15s or older is eligible for compactions
+		SizeBuckets:                DefaultSizeBuckets,                 // sizes defined above
+		OrderBy:                    TasksOrderedByOldestMutableAndSize, // compact mutable segments first
+	}
+)
+
+// NewPlan returns a new compaction.Plan per the rules above and the knobs provided.
+func NewPlan(candidateSegments []Segment, opts PlannerOptions) (*Plan, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	plan := &Plan{
+		OrderBy: opts.OrderBy,
+	}
+	// Planning is a two-phase process:
+	// (1) Identify all "compactable" Segments:
+	//    - All mutable segments are compactable (over age Y)
+	//    - All immutable segments (below size MaxCompactionSize) are compactable
+	//
+	// (2) Come up with a logical plan for compactable segments
+	//    (a) Group the segments into given buckets (compactions can only be performed for segments within the same bucket)
+	//    (b) For each bucket:
+	//    (b1) Accumulate segments until cumulative size is over the max of the current bucket.
+	//    (b2) Add a Task which comprises segments from (b1) to the Plan.
+	//    (b3) Continue (b1) until the bucket is empty.
+	//    (c) Priotize Tasks w/ "compactable" Mutable Segments over all others
+
+	// 1st phase - find all compactable segments
+	var compactableSegments []Segment
+	for _, seg := range candidateSegments {
+		compactable := (seg.Type == segments.FSTType && seg.Size < opts.MaxImmutableCompactionSize) ||
+			(seg.Type == segments.MutableType &&
+				(seg.Age >= opts.MutableCompactionAge || seg.Size >= opts.MaxMutableSegmentSize))
+		if compactable {
+			compactableSegments = append(compactableSegments, seg)
+		} else {
+			plan.UnusedSegments = append(plan.UnusedSegments, seg)
```

Comment:
Hm should probably also preallocate the UnusedSegments slice to have capacity `len(candidateSegments) too.

Reply:
sure, i'll lazily allocate

Reply:
done